File: 백준/Gold/1005. ACM Craft/ACM Craft.py
# https://www.acmicpc.net/problem/1005
t = int(input())
for _ in range(t):
    N, K = map(int, input().split())
    adj = [[] for _ in range(N + 1)]
    ind = [0 for _ in range(N + 1)]
    dp = [0 for _ in range(N + 1)]
    times = ['인덱스용'] + list(map(int, input().split()))
    for _ in range(K):
        X, Y = map(int, input().split())
        adj[X].append(Y)
        ind[Y] += 1
    W = int(input())
    # 후보군 넣기
    candidates = []
    for num in range(1, N + 1):
        if ind[num] == 0:
            candidates.append(num)
            dp[num] = times[num]
    # candidates 확인  
    result = []
    while candidates:
        cur = candidates.pop()
        result.append(cur)
        
        for nxt in adj[cur]:
            ind[nxt] -= 1
            # times[nxt] 반드시 더해져야함. 어디에? 내 현재 누적 시간에다가.
            # dp[nxt]가 갱신되어야함. 
            dp[nxt] = max(dp[nxt], times[nxt] + dp[cur])
            if ind[nxt] == 0:
                candidates.append(nxt)

    print(dp[W])

File: 백준/Gold/1005. ACM Craft/ACM Craft.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        for (int i = 0; i < t; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken());
            int K = Integer.parseInt(st.nextToken());

            ArrayList<Integer>[] adj = new ArrayList[N + 1];
            int[] ind = new int[N + 1];
            int[] dp = new int[N + 1];
            int[] times = new int[N + 1];

            StringTokenizer timeTokens = new StringTokenizer(br.readLine());
            for (int j = 1; j <= N; j++) {
                times[j] = Integer.parseInt(timeTokens.nextToken());
                adj[j] = new ArrayList<>();
            }

            for (int j = 0; j < K; j++) {
                st = new StringTokenizer(br.readLine());
                int X = Integer.parseInt(st.nextToken());
                int Y = Integer.parseInt(st.nextToken());
                adj[X].add(Y);
                ind[Y]++;
            }

            int W = Integer.parseInt(br.readLine());
            Queue<Integer> candidates = new LinkedList<>();

            for (int num = 1; num <= N; num++) {
                if (ind[num] == 0) {
                    candidates.add(num);
                    dp[num] = times[num];
                }
            }

            ArrayList<Integer> result = new ArrayList<>();
            while (!candidates.isEmpty()) {
                int cur = candidates.poll();
                result.add(cur);

                for (int nxt : adj[cur]) {
                    ind[nxt]--;
                    dp[nxt] = Math.max(dp[nxt], times[nxt] + dp[cur]);

                    if (ind[nxt] == 0) {
                        candidates.add(nxt);
                    }
                }
            }

            System.out.println(dp[W]);
        }
    }
}
